src/os/bluestore/BlueStore.cc

```diff
@@ -7036,14 +7146,996 @@ int BlueStore::_mount()
+  uint32_t pad[0x2];		// 0x28
+
+  allocator_image_trailer(const allocator_image_header *p_header, uint32_t entries_count) {
+    memset((char*)&(this->null_extent), 0, sizeof(this->null_extent));
+    this->timestamp       = p_header->timestamp;
+    this->format_version  = p_header->format_version;
+    this->valid_signature = ALLOCATOR_IMAGE_VALID_SIGNATURE;
+    this->serial          = p_header->serial;
+    this->entries_count   = entries_count;
+    memset(this->pad, 0, sizeof(this->pad));
+  }
+};
+
+//-------------------------------------------------------------------------------------
+int BlueStore::invalidate_allocation_file_on_bluestore()
+{
+  BlueFS::FileWriter *p_handle = nullptr;
+  int ret = 0;
+
+  dout(5) << "==========================================================" << dendl;
+  dout(5) << "invalidate_allocation_file" << dendl;
+    
+  if (!bluefs->dir_exists(allocator_dir)) {
+    derr << "Failed mkdir with error-code " << ret << dendl;
+    return -1;
+  }
+  
+  ret = bluefs->open_for_write(allocator_dir, allocator_file, &p_handle, false);
+  if (ret != 0) {
+    derr << "Failed open_for_write with error-code " << ret << dendl;
+    return -1;
+  }
+
+  bluefs->truncate(p_handle, 0);
+  bluefs->fsync(p_handle);
+  bluefs->close_writer(p_handle);
+
+  return 0;
+}
+
+static const uint64_t header_count_in_extents  = sizeof(allocator_image_header)  / sizeof(extent_t);
+static const uint64_t trailer_count_in_extents = sizeof(allocator_image_trailer) / sizeof(extent_t);
+const unsigned MAX_EXTENTS_IN_BUFFER = 16 * 1024; // 16K extents
+
+// write the allocator to a flat bluefs file - 16k extents at a time
+//-----------------------------------------------------------------------------------
+int BlueStore::store_allocator(Allocator* allocator)
+{
+  extent_t buffer[MAX_EXTENTS_IN_BUFFER + trailer_count_in_extents]; // 192KB
+  uint64_t file_offset = 0;
+  static_assert(sizeof(allocator_image_header) == sizeof(allocator_image_trailer));
+  static_assert(sizeof(allocator_image_header) %  sizeof(extent_t) == 0);
+  //static_assert(sizeof(allocator_image_header) == 0x30 );  
+
+  uint64_t stored_alloc_size = 0;
+  utime_t  start_time = ceph_clock_now();
+  BlueFS::FileWriter *p_handle = nullptr;
+  int ret = 0;
+  
+  if (!bluefs->dir_exists(allocator_dir) ) {
+    ret = bluefs->mkdir(allocator_dir);
+    if (ret != 0) {
+      derr << "Failed mkdir with error-code " << ret << dendl;
+      return -1;
+    }
+  }
+
+  // reuse previous file-allocation is exists
+  ret = bluefs->open_for_write(allocator_dir, allocator_file, &p_handle, false);
+  if (ret != 0) {
+    derr << "Failed open_for_write with error-code " << ret << dendl;
+    return -1;
+  }
+
+  uint64_t        extent_count  = 0;
+  extent_t       *p_curr        = buffer;
+  const extent_t *p_end         = buffer + MAX_EXTENTS_IN_BUFFER;
+  allocator_image_header  header(s_format_version, s_serial);
+  memcpy((byte*)p_curr, (byte*)&header, sizeof(header));
```

Comment:
rather than copy the raw struct, allocator_image_header should implement a DENC() method that calls denc() on each relevant data member - see bluestore_onode_t::DENC() for an example, as well as include/denc.h.

This makes the format compatible with different endianness (helpful for analyzing a disk image from a different architecture) and allows us to easily version it if we need to add more fields later.